packages/http-crawler/src/internals/file-download.ts

```diff
@@ -58,6 +59,85 @@ export type FileDownloadRequestHandler<
     JSONData extends Dictionary = any, // with default to Dictionary we cant use a typed router in untyped crawler
 > = RequestHandler<FileDownloadCrawlingContext<UserData, JSONData>>;
 
+/**
+ * Creates a transform stream that throws an error if the source data speed is below the specified minimum speed.
+ * This `Transform` checks the amount of data every `checkProgressInterval` milliseconds.
+ * If the stream has received less than `minSpeedKbps * historyLengthMs / 1000` bytes in the last `historyLengthMs` milliseconds,
+ * it will throw an error.
+ *
+ * Can be used e.g. to abort a download if the network speed is too slow.
+ * @returns Transform stream that monitors the speed of the incoming data.
+ */
+export function MinimumSpeedStream({
+    minSpeedKbps,
+    historyLengthMs = 10e3,
+    checkProgressInterval: checkProgressIntervalMs = 5e3,
+}: {
+    minSpeedKbps: number;
+    historyLengthMs?: number;
+    checkProgressInterval?: number;
+}): Transform {
+    let snapshots: { timestamp: number; bytes: number }[] = [];
+
+    const checkInterval = setInterval(() => {
+        const now = Date.now();
+
+        snapshots = snapshots.filter((snapshot) => now - snapshot.timestamp < historyLengthMs);
+        const totalBytes = snapshots.reduce((acc, snapshot) => acc + snapshot.bytes, 0);
+        const elapsed = (now - (snapshots[0]?.timestamp ?? 0)) / 1000;
+
+        if (totalBytes / 1024 / elapsed < minSpeedKbps) {
+            clearInterval(checkInterval);
+            stream.emit('error', new Error(`Stream speed too slow, aborting...`));
+        }
+    }, checkProgressIntervalMs);
+
+    const stream = new Transform({
+        transform: (chunk, _, callback) => {
+            snapshots.push({ timestamp: Date.now(), bytes: chunk.length });
+            callback(null, chunk);
+        },
+        final: (callback) => {
+            clearInterval(checkInterval);
+            callback();
+        },
+    });
+
+    return stream;
+}
+
+/**
+ * Creates a transform stream that logs the progress of the incoming data.
+ * This `Transform` calls the `logProgress` function every `loggingInterval` milliseconds with the number of bytes received so far.
+ *
+ * Can be used e.g. to log the progress of a download.
+ * @returns Transform stream logging the progress of the incoming data.
+ */
+export function ProgressLoggerStream({
+    logProgress,
+    loggingInterval = 5000,
+}: { logProgress: (transferredBytes: number) => void; loggingInterval?: number }): Transform {
```

Comment:
Is there any way we could also give this the total length of the file, if it's available?

Reply:
As discussed in person, the helper accepts arbitrary streams, so the semantics of file download (total file size etc.) aren't always right. I renamed the function to `ByteCounterStream` to reflect this better.

Note that you can still use this function to log the file download progress by e.g. storing the total file download in closure with the `logTransferredBytes` function.